students/mattmaeda/session05/comprehensions.py

```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env python3
+
+# Problems from
+# https://uwpce-pythoncert.github.io/PythonCertDevel/exercises/comprehensions_lab.html#exercise-comprehensions
+
+def count_evens(num_list):
+    # http://codingbat.com/prob/p189616
+    # Using a list comprehension, return the number of even integers in the
+    # given list.
+    return len([i for i in num_list if i % 2 == 0])
+
+
+def dict_formatter(ref_dict):
+    # dict/set problem 1
+    #
+    # Print the dict by passing it to a string format method, so that you get
+    # something like:
+    #
+    # “Chris is from Seattle, and he likes chocolate cake, mango fruit, greek
+    # salad, and lasagna pasta”
+    return "%(name)s is from %(city)s, and he likes %(cake)s cake, %(fruit)s " \
+           "fruit, %(salad)s salad and %(pasta)s pasta" % {k: v for k, v in ref_dict.items()}
+
+
+def int_to_hex_mapping(n):
+    # dict/set problems 2 and 3
+    #
+    # Using a list comprehension, build a dictionary of numbers from zero to
+    # fifteen and the hexadecimal equivalent (string is fine). (the hex()
+    # function gives you the hexidecimal representation of a number as a string)
+    #
+    # Do the previous entirely with a dict comprehension – should be a one-liner
+    return {i: hex(i)for i in range(n+1)}
+
+
+def count_letter_a_in_dict(ref_dict):
+    # dict/set problem 4
+    #
+    # Using the dictionary from item (1): Make a dictionary using the same keys
+    # but with the number of ‘a’s in each value. You can do this either by
+    # editing the dict in place, or making a new one. If you edit in place make
+    # a copy first!
+    return {k: v.count("a") for k,v in ref_dict.items()}
+
+
+def set_comprehensions(divisors, max_range):
+    # dict/set problem 5
+    #
+    # Create sets s2, s3 and s4 that contain numbers from zero through twenty,
+    # divisible 2, 3 and 4.
+    #
+    #   a. Do this with one set comprehension for each set.
+    #   b. What if you had a lot more than 3? – Don’t Repeat Yourself (DRY).
+    #       1. create a sequence that holds all the divisors you might want –
+    #          could be 2,3,4, or could be any other arbitrary divisors.
+    #       2. loop through that sequence to build the sets up – so no repeated
+    #          code. you will end up with a list of sets – one set for each
+    #          divisor in your sequence.
+    #       3. The idea here is that when you see three (Or more!) lines of code
+    #          that are almost identical, then you you want to find a way to
+    #          generalize that code and have it act on a set of inputs, so the
+    #          actual code is only written once.
+    #   c. Extra credit: do it all as a one-liner by nesting a set comprehension
+    #      inside a list comprehension. (OK, that may be getting carried away!)
+    return [{i for i in range(max_range+1) if i % d == 0} for d in divisors]
+
```

Comment:
Great work! Note that for integers, 0 is False and other numbers are True, so you can replace `if i % d == 0` with `if not i % d` Not much shorter, nor necessarily more readable, but it does get used, so worth being familiar with. I think computationally the if not idiom is slightly faster, but rarely would that matter.